File: 1_gcd/1_gcd.py
def listing_all(A:int, B:int)-> int:
    fA = []
    fB = []
    fC = []

    for i in range(1,A+1):
        if A%i == 0: fA.append(i)

    for i in range(1,B+1):
        if B%i == 0: fB.append(i)

    for element in fA:
        if element in fB:
            fC.append(element)
    return fC[-1]

def finding_common(A:int, B:int)-> int:
    print('finding_common')
    fC = []
    for i in range(1,A+1):
        if A%i == 0 and B%i==0: fC.append(i)
    return fC[-1]

def euclid(A:int, B:int)-> int:
    # Using this lemma, we express the larger number “a” in terms of the smaller number “b” in the quotient remainder form a=bq+r
    # if r=0 b is gcd and r = a%b
    print('euclid')
    while B != 0:
        A, B = B, A % B

    return abs(A)

if __name__ == '__main__':
    A, B = list(map(int,input().split()))
    # print(listing_all(A,B))
    # print(finding_common(A,B))
    print(euclid(A,B))

File: 1_gcd/1_gcd.cpp
#include <iostream>
#include <vector>
#include <algorithm>

// Function to list all factors of A and B and return the last common factor (GCD)
std::vector<int> listing_all(int A, int B) {
    std::vector<int> fA;
    std::vector<int> fB;
    std::vector<int> fC;

    for (int i = 1; i <= A; ++i) {
        if (A % i == 0) fA.push_back(i);
    }

    for (int i = 1; i <= B; ++i) {
        if (B % i == 0) fB.push_back(i);
    }

    for (const auto& element : fA) {
        if (std::find(fB.begin(), fB.end(), element) != fB.end()) {
            fC.push_back(element);
        }
    }

    return fC;
}

// Function to find common factors of A and B and return the last common factor (GCD)
int finding_common(int A, int B) {
    std::cout << "finding_common" << std::endl;
    std::vector<int> fC;

    for (int i = 1; i <= A; ++i) {
        if (A % i == 0 && B % i == 0) {
            fC.push_back(i);
        }
    }

    return fC.back();
}

// Function to calculate GCD using Euclid's algorithm
int euclid(int A, int B) {
    std::cout << "euclid" << std::endl;
    while (B != 0) {
        int temp = B;
        B = A % B;
        A = temp;
    }

    return abs(A);
}

int main() {
    int A, B;
    std::cin >> A >> B;

    // std::vector<int> result = listing_all(A, B);
    // std::cout << result.back() << std::endl;

    // std::cout << finding_common(A, B) << std::endl;

    std::cout << euclid(A, B) << std::endl;

    return 0;
}